File: python/src/skillflag/install/copy.py
"""Skill directory copy with conflict handling (port of install/copy.ts)."""

from __future__ import annotations

import os
import shutil

from skillflag.install.errors import InstallError


def copy_skill_dir(source_dir: str, dest_dir: str, force: bool) -> None:
    if os.path.exists(dest_dir):
        if not force:
            raise InstallError(f"Destination already exists: {dest_dir}")
        if os.path.isdir(dest_dir) and not os.path.islink(dest_dir):
            shutil.rmtree(dest_dir)
        else:
            os.remove(dest_dir)

    parent = os.path.dirname(dest_dir)
    if parent:
        os.makedirs(parent, exist_ok=True)
    # copy2 preserves file modes, so execute bits survive the copy.
    shutil.copytree(source_dir, dest_dir)

File: rust/crates/skillflag/src/install/copy.rs
use std::fs;
use std::path::Path;

use crate::core::errors::{Result, SkillflagError};

fn copy_tree(source: &Path, dest: &Path) -> Result<()> {
    fs::create_dir_all(dest)?;
    for entry in fs::read_dir(source)? {
        let entry = entry?;
        let file_type = entry.file_type()?;
        let target = dest.join(entry.file_name());
        if file_type.is_dir() {
            copy_tree(&entry.path(), &target)?;
        } else if file_type.is_symlink() {
            #[cfg(unix)]
            {
                let link_target = fs::read_link(entry.path())?;
                std::os::unix::fs::symlink(link_target, &target)?;
            }
            #[cfg(not(unix))]
            {
                fs::copy(entry.path(), &target)?;
            }
        } else {
            // fs::copy preserves permissions (and therefore execute bits).
            fs::copy(entry.path(), &target)?;
        }
    }
    Ok(())
}

/// Copy a skill directory into place. Fails when the destination exists
/// unless `force`, in which case the destination is removed and re-copied.
pub fn copy_skill_dir(source_dir: &Path, dest_dir: &Path, force: bool) -> Result<()> {
    if let Ok(metadata) = fs::symlink_metadata(dest_dir) {
        if !force {
            return Err(SkillflagError::new(format!(
                "Destination already exists: {}",
                dest_dir.display()
            )));
        }
        if metadata.is_dir() {
            fs::remove_dir_all(dest_dir)?;
        } else {
            fs::remove_file(dest_dir)?;
        }
    }

    if let Some(parent) = dest_dir.parent() {
        fs::create_dir_all(parent)?;
    }
    copy_tree(source_dir, dest_dir)
}
